src/snowflake/snowpark/mock/_stored_procedure.py

```diff
@@ -0,0 +1,238 @@
+from snowflake.snowpark.dataframe import DataFrame
+from snowflake.snowpark.exceptions import SnowparkSQLException
+from snowflake.snowpark.mock import CUSTOM_JSON_ENCODER
+from snowflake.snowpark.mock._plan import calculate_expression
+from snowflake.snowpark.mock._snowflake_data_type import ColumnEmulator
+from snowflake.snowpark.stored_procedure import (
+    StoredProcedure,
+    StoredProcedureRegistration,
+)
+from snowflake.snowpark.types import (
+    ArrayType,
+    DataType,
+    MapType,
+    StructType,
+    _FractionalType,
+    _IntegralType,
+)
+
+from ._telemetry import LocalTestOOBTelemetryService
+
+if sys.version_info <= (3, 9):
+    from typing import Iterable
+else:
+    from collections.abc import Iterable
+
+
+def sproc_types_are_compatible(x, y):
+    if (
+        isinstance(x, type(y))
+        or isinstance(x, _IntegralType)
+        and isinstance(y, _IntegralType)
+        or isinstance(x, _FractionalType)
+        and isinstance(y, _FractionalType)
+    ):
+        return True
+    return False
+
+
+class MockStoredProcedure(StoredProcedure):
+    def __call__(
+        self,
+        *args: Any,
+        session: Optional["snowflake.snowpark.session.Session"] = None,
+        statement_params: Optional[Dict[str, str]] = None,
+    ) -> Any:
+        args, session = self._validate_call(args, session)
+
+        # Unpack columns if passed
+        parsed_args = []
+        for arg, expected_type in zip(args, self._input_types):
+            if isinstance(arg, Column):
+                expr = arg._expression
+
+                # If expression does not define its datatype we cannot verify it's compatibale.
+                # This is potentially unsafe.
+                if expr.datatype and not sproc_types_are_compatible(
+                    expr.datatype, expected_type
+                ):
+                    raise ValueError(
+                        f"Unexpected type {expr.datatype} for sproc argument of type {expected_type}"
+                    )
+
+                # Expression may be a nested expression. Expression should not need any input data
+                # and should only return one value so that it can be passed as a literal value.
+                # We pass in a single None value so that the expression evaluator has some data to
+                # pass to the expressions.
+                resolved_expr = calculate_expression(
+                    expr,
+                    ColumnEmulator(data=[None]),
+                    session._analyzer,
+                    {},
+                )
+
+                # If the length of the resolved expression is not a single value we cannot pass it as a literal.
+                if len(resolved_expr) != 1:
+                    raise ValueError(
+                        "[Local Testing] Unexpected argument type {expr.__class__.__name__} for call to sproc"
+                    )
+                parsed_args.append(resolved_expr[0])
```

Comment:
can you give me an example of stored procedure with one `arg` being Column? I'm trying to understand how this arg is used -- is it used to getting column of data within this process or we just need to parse the expression into a literal.

if this is just analyzing an expression to infer a literal value, then the MockAnalyzer.analyze can better help here: https://github.com/snowflakedb/snowpark-python/blob/main/src/snowflake/snowpark/mock/_analyzer.py#L169-L176.

Reply:
This test has an example of where this is used:
https://github.com/snowflakedb/snowpark-python/pull/1338/files#diff-4963488bfdaad538e6af7222d4db977758e66d05b4c11b600583449152416f50L221

Reply:
I see, if evaluating the literal value is the only case, then let's try using `session._analyzer.analyze`.

calculate_expression is for calculating table/column results, while analyze is just for expressions parsing

Reply:
Running analyze on this statement produces query text which is not what the function expects as input. For example, the sqrt done in the test when put through the analyzer creates the string `sqrt(36)` which is not what the function expects as input.

Reply:
I see, that makes sense. thanks for checking it!